filters/include/pcl/filters/experimental/transform_filter.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2020-, Open Perception
+ *
+ *  All rights reserved
+ */
+
+#pragma once
+
+#include <pcl/common/point_tests.h>
+#include <pcl/filters/filter.h>
+
+namespace pcl {
+namespace experimental {
+
+/**
+ * \brief The base of GridFilterBase which has the common member functions and
+ * attribute for the grid filters.
+ * \ingroup filters
+ */
+template <typename GridStruct, typename PointT>
+class TransformFilter : public Filter<PointT>, public GridStruct {
+protected:
+  using Filter<PointT>::filter_name_;
+  using Filter<PointT>::getClassName;
+  using Filter<PointT>::input_;
+
+  using PointCloud = typename Filter<PointT>::PointCloud;
+  using PointCloudPtr = typename PointCloud::Ptr;
+  using PointCloudConstPtr = typename PointCloud::ConstPtr;
+
+public:
+  /** \brief Empty constructor. */
+  TransformFilter()
+  {
+    if (GridStruct::filter_name_.empty()) {
+      PCL_WARN("[pcl::%s] Filter name is empty\n", getClassName().c_str());
+    }
+    filter_name_ = GridStruct::filter_name_;
+  }
+
+  /** \brief Destructor. */
+  ~TransformFilter() {}
+
+protected:
+  /** \brief Filter a Point Cloud based on the templated grid structure
+   * \param[out] output the resultant point cloud message
+   */
+  void
+  applyFilter(PointCloud& output) override
+  {
+    // Has the input dataset been set already?
+    if (!input_) {
+      PCL_WARN("[pcl::%s::applyFilter] No input dataset given!\n",
+               getClassName().c_str());
+      return;
+    }
+
+    output.height = 1;      // downsampling breaks the organized structure
+    output.is_dense = true; // we filter out invalid points
+
+    if (!GridStruct::setUp()) {
+      output = *input_;
+      return;
+    }
+
+    for (const auto& pt : *input_) {
+      if (input_->is_dense || isXYZFinite(pt)) {
```

Comment:
Good point, I try it out when I start benchmarking. But I found this checking is also performed by `PassThrough` filter, we can also try call `PassThrough` ? It would looks more concise

Reply:
FYI: I found for all grid filters expect `ApproximateVoxelGrid` do this

Reply:
I benchmarked both way, but the benchmark shows having a loop is faster, which is quite weird for me...

## One loop
```cpp
for (const auto& pt : *input_) {
      if (input_->is_dense || isXYZFinite(pt)) {
        GridStruct::addPointToGrid(pt);
      }
    }
```
```
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
BM_VoxelGrid_milk                    10.8 ms         10.8 ms           65
BM_ApproximateVoxelGrid_milk         7.49 ms         7.47 ms           91
BM_Experimental_VoxelGrid_milk       8.31 ms         8.28 ms           84
BM_VoxelGrid_mug                     8.53 ms         8.50 ms           82
BM_ApproximateVoxelGrid_mug          6.81 ms         6.79 ms          101
BM_Experimental_VoxelGrid_mug        6.28 ms         6.26 ms          110
```

## Two loops
```cpp
    if (input_->is_dense) {
      for (const auto& pt : *input_)
        GridStruct::addPointToGrid(pt);
    }
    else {
      for (const auto& pt : *input_) {
        if (isXYZFinite(pt)){
          GridStruct::addPointToGrid(pt);
        }
      }
    }
```
```
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
BM_VoxelGrid_milk                    10.7 ms         10.7 ms           64
BM_ApproximateVoxelGrid_milk         7.40 ms         7.40 ms           89
BM_Experimental_VoxelGrid_milk       9.04 ms         9.03 ms           79
BM_VoxelGrid_mug                     8.67 ms         8.66 ms           80
BM_ApproximateVoxelGrid_mug          6.80 ms         6.79 ms          102
BM_Experimental_VoxelGrid_mug        7.04 ms         7.04 ms          101
```

Reply:
Yes, that is strange. The mug and milk dataset are not dense when the filter receives them, right? It would be interesting to also do the benchmarking with dense clouds because in that case the two loops should really be faster than the one loop